AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1282,19 +1281,51 @@ - (void)_tearDownPlaceholderLayer
 
 #pragma mark - For Subclasses
 
+- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  if (_methodOverrides & ASDisplayNodeMethodOverrideCalculateSizeThatFits) {
+    CGSize size = [self calculateSizeThatFits:constrainedSize.max];
+    return [ASLayout newWithLayoutableObject:self size:ASSizeRangeClamp(constrainedSize, size)];
+  } else {
+    id<ASLayoutable> layoutSpec = [self layoutSpecThatFits:constrainedSize];
+    ASLayout *layout = [layoutSpec measureWithSizeRange:constrainedSize];
+    // Make sure layoutableObject of the root layout is `self`, so that the flattened layout will be structurally correct.
+    if (layout.layoutableObject != self) {
+      layout.position = CGPointZero;
```

Comment:
This condition actually will happen most (if not all) of the time people if a layout spec is returned (via -layoutSpecThatFits:). For example KittenNode returns a layout spec with ASInsetLayoutSpec as layoutable object of the root layout.

There are 2 solutions for this problem:
- -layoutSpecThatFits: requires users to wrap his layout spec inside an umbrella spec that have valid root layoutable object.
- Automatically wrap the calculated layout, which is what I'm doing.